### ISSUE TITLE
Inconsistency in learner API?

### ISSUE THREAD

**Issue by larskotthoff** [mlr-org/mlr3#285]
What's the reasoning behind the APIs for train and predict for learners? Train modifies the learner internally by storing the model and returns the result only invisibly -- so it's not supposed to be used in a functional way? Predict on the other hand makes no internal modifications and returns the result. It seems counter-intuitive that those two closely-related functions behave differently in this respect.

This can lead to some weird situations; for example `mlr_learners$get("classif.rpart")$train(mlr_tasks$get("iris")` runs without warnings or errors, but does absolutely nothing.

**Comment by mllg** [comment]
* `train()` returns itself, not the result. This operation mutates the learner, i.e. stores the model. You can now call `$importance()` or similar getters to extract information from the model. 

* `predict()` returns a prediction object using the stored model. You might want to call this method multiple times, on different test sets or new data. 

I agree that this behaviour seems odd if you come from mlr-v2 or caret. sklearn is quite similar in the behaviour of `train`/`fit` and `predict` though.

The above line is valid code, and if you assign it, you should get a trained learner:
```r
lrn = mlr_learners$get("classif.rpart")$train(mlr_tasks$get("iris")
lrn$model
```

Does this help?

**Comment by larskotthoff** [comment]
Well, you might also want to train a model multiple times with different data. I tend to find things like this counter-intuitive and don't like hidden state in general -- it's easy to get confused and accidentally modify something you didn't mean to modify.

What's the reason for choosing this behaviour and not the old mlr2-style?

**Comment by berndbischl** [comment]
please keep the comments coming but i dont think this specific thing will be changed.

it has the advantage of "unifying" the learner API now in one class. some methods refer to the model, some to the learner. the use can now implement them in one class.

eg the training vs the "get_feature_importance" methods and so on

before you say "hey, ok, but one should not choose a weird design just to obtain some shortcuts!"

i also believe the design is somewhat "natural".

see it like this:
the learner encode the "prediction function" - as a mathematical object.
before you can apply this, you need to set its parameters. which you indirectly by passing data to the training and them optimizing for the parameters. 
so from that viewpoint it really IS one class.

and we do excatly the same approach in pipelines. and sklearn seems to agree.
for me that is enough reasons.

**Comment by berndbischl** [comment]
> Well, you might also want to train a model multiple times with different data.

and that is sure possible. simply clone the learner k times, then train of k different tasks.
or overwrite a currently trained state
